lambdas/es/indexer/index.py

```diff
@@ -299,6 +304,94 @@ def do_index(
         logger_.debug("%s indexed as package (%s)", key, event_type)
 
 
+def _try_parse_date(s: str) -> Optional[datetime.datetime]:
+    # XXX: do we need to support more formats?
+    if s[-1:] == "Z":
+        s = s[:-1]
+    try:
+        return datetime.datetime.fromisoformat(s)
+    except ValueError:
+        return None
+
+
+MAX_KEYWORD_LEN = 256
+
+
+def _get_metadata_fields(path: tuple, d: dict):
+    for k, raw_value in d.items():
+        if isinstance(raw_value, dict):
+            yield from _get_metadata_fields(path + (k,), raw_value)
+        else:
+            v = raw_value
+            if isinstance(v, str):
+                date = _try_parse_date(v)
+                if date is not None:
+                    type_ = "date"
+                    v = date
+                else:
+                    type_ = "keyword" if len(v) <= MAX_KEYWORD_LEN else "text"
+            elif isinstance(v, bool):
+                type_ = "boolean"
+            elif isinstance(v, (int, float)):
+                # XXX: do something on ints that can't be converted to float without loss?
+                type_ = "double"
+            elif isinstance(v, list):
+                if not (v and all(isinstance(x, str) for x in v)):
+                    continue
+                type_ = "keyword" if all(len(x) <= MAX_KEYWORD_LEN for x in v) else "text"
+            else:
+                logger.warning("ignoring value of type %s", type(v))
+                continue
+
+            yield path + (k,), type_, raw_value, v
+
+
+def get_metadata_fields(meta):
+    if not isinstance(meta, dict):
+        # XXX: can we do something better?
+        return None
+    return [
+        {
+            "json_pointer": jsonpointer.JsonPointer.from_parts(path).path,
+            "type": type_,
+            "text": json.dumps(raw_value, ensure_ascii=False),
+            type_: value,
+        }
+        for path, type_, raw_value, value in _get_metadata_fields((), meta)
+    ]
+
+
+def _prepare_workflow_for_es(workflow, bucket):
+    if workflow is None:
+        return None
+
+    try:
+        config_url = workflow["config"]
+        if not config_url.startswith(f"s3://{bucket}/.quilt/workflows/config.yml"):
+            raise Exception(f"Bad workflow config URL {config_url}")
+
+        config_url_parsed = urllib.parse.urlparse(config_url)
+        query = urllib.parse.parse_qs(config_url_parsed.query)
+        version_id = query.pop('versionId', [None])[0]
+        if query:
+            raise Exception(f"Unexpected S3 query string: {config_url_parsed.query!r}")
+
+        return {
+            "config_version_id": version_id,  # XXX: how to handle None?
```

Comment:
well i guess None is a valid case (non-versioned buckets), so we just use it as is. or is it "null" string? which is also okay, probably, untils it works as expected with s3 apis. anyways, it's not used in the current version of faceted search, so doesn't matter that much